### PR TITLE
[8.0] Replace __del__ with weakref.finalize in DictCache

### PR DESCRIPTION
**Pull request by chrisburr**
See https://github.com/DIRACGrid/DIRAC/pull/7653 and https://docs.python.org/3/library/weakref.html#comparing-finalizers-with-del-methods

BEGINRELEASENOTES

*Core
FIX: Replace `__del__` with weakref.finalize in DictCache

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/Core/Utilities/DictCache.py [hunk 1]
```diff
@@ -64,6 +65,8 @@ def __init__(self, deleteFunction=False, threadLocal=False):
         # Function to clean the elements
         self.__deleteFunction = deleteFunction
 
+        self.__finalizer = weakref.finalize(self, self.purgeAll, useLock=False)
```
Comment (chrisburr) [inline src/DIRAC/Core/Utilities/DictCache.py]: The use of `purgeAll` here should probably be refactored to not depend on self and go pass in the members it needs instead.

**Comment by chaen** [comment]
Ah cool, I did not know about `weakref`, and apparently, it's the right thing to do :laughing: https://github.com/cupy/cupy/pull/8315

File: src/DIRAC/Core/Utilities/DictCache.py (hunk 1 above)
Comment (chaen) [reply src/DIRAC/Core/Utilities/DictCache.py]: Fine by me, do you want to add it in this PR ?

Comment (chrisburr) [reply src/DIRAC/Core/Utilities/DictCache.py]: Done

**Review by chaen — APPROVED** [review]

**Pull request merged**